Source/Session.swift

```diff
@@ -552,23 +585,48 @@ extension Session: RequestDelegate {
         }
     }
 
-    public func cancelRequest(_ request: Request) {
-        rootQueue.async {
-            request.didCancel()
+    public func attemptToTransition(_ request: Request, to state: Request.State) {
+        requestStateQueue.sync {
+            guard request.protectedMutableState.attemptToTransitionTo(state) else { return }
```

Comment:
All of this logic can be redesigned to be MUCH cleaner by combining state transition logic with the `requestStateQueue` itself in some specialized APIs. This is just a first attempt to show how it could actually work, then we can clean it up if we like it.

Reply:
Refactored in 613f6314 into a `finishRequestIfCancelled(_:)` API. It cleans up the logic a TON, but I'm not in love with the name. It was the best I could come up with.